text/0000-floating-point-additional-precision.md

```diff
@@ -0,0 +1,141 @@
+mathematically exact answer. The implementation of Rust for any target platform
+must provide at least that much accuracy. In some cases, Rust can perform
+operations with higher accuracy than required, and doing so provides greater
+performance (such as by removing intermediate rounding steps). This will never
+make any floating-point operation *less* accurate, but it can make
+floating-point operations *more* accurate, making the result closer to the
+mathematically exact answer.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Currently, Rust's [specification for floating-point
+types](https://doc.rust-lang.org/reference/types/numeric.html#floating-point-types)
+states that:
+> The IEEE 754-2008 "binary32" and "binary64" floating-point types are f32 and f64, respectively.
+
+This RFC proposes updating that definition as follows:
+
+The `f32` and `f64` types represent the IEEE 754-2008 "binary32" and "binary64"
+floating-point types. Operations on those types must provide no less
+precision than the IEEE standard requires; such operations may provide *more*
+precision than the standard requires, such as by doing a series of operations
+with higher precision before storing a value of the desired precision.
+
+rustc may provide a codegen (`-C`) option to disable this behavior, such as `-C
+disable-extra-fp-precision`. Rust may also provide an attribute to disable this
+behavior from within code, such as `#[disable_extra_fp_precision]`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If Rust already provided bit-for-bit identical floating-point computations
+across platforms, this change could potentially allow floating-point
+computations to differ by platform (though never below the standards-required
+accuracy). However, standards-compliant implementations of math functions on
+floating-point values may already vary slightly by platform, sufficiently so to
+produce different binary results. This proposal can never make results *less*
+accurate, it can only make results *more* accurate.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could provide a separate set of types and allow extra accuracy in their
+operations; however, this would create ABI differences between floating-point
+functions, and the longer, less-well-known types seem unlikely to see
+widespread use.
+
+We could provide an option to enable extra accuracy for the default
+floating-point types, but disable it by default. This would leave the majority
+of floating-point code unable to use these optimizations, however; defaults
+matter, and the majority of code seems likely to use the defaults.
+
+We could do nothing, and require code to use `a.mul_add(b, c)` for
+optimization; however, this would not allow for similar future optimizations,
+and would not allow code to easily enable this optimization without substantial
+code changes.
```

Comment:
We could provide a clippy lint that recognizes `a * b + c` (and many others), and tell people that if they don't care about precision, they can write `a.mul_add(b, c)` instead. We could have a group of clippy lints about these kind of things that people can enable in bulk.

Reply:
On this particular point a clippy lint is helpful but not necessarily enough. Once the optimizer chews through layers of code it can end up at an `a * b + c` expression without it being anything that is obvious to clippy.